Master datagrid row overlay

stather opened the pull request:
Adds 2 new event callback to datagrid, and a feature to display an overlay when the mouse is over the row.

New Parameter RowOver of type EventCallback<DataGridRowMouseEventArgs<TItem>>
This is called when the mouse is over the row.

New Parameter RowLeave of type EventCallback<DataGridRowMouseEventArgs<TItem>>
This is called when the mouse leaves a row.

These are useful anyway, but used mainly to facilitate the row overlay feature.

New Parameter OverlayTemplate of type RenderFragment<TItem>.

This render fragment can be used to display additional information about the row item when the mouse is over the row. The content of the fragment is displayed over the top of the current row and is right aligned.

```
<DataGrid TItem="sometype">
                       <OverlayTemplate>
                            <Div Flex="Flex.JustifyContent.End.AlignItems.Center" Margin="Margin.Is3.FromBottom" Style="width:400px;background-color:white;">
                                <Div Padding="Padding.Is2">
                                    <Button Size="Size.Small" Color="Color.Secondary" onclick="event.stopPropagation();">Ack</Button>
                                </Div>
                                <Div Padding="Padding.Is2">
                                    <Button Size="Size.Small" Color="Color.Secondary" onclick="event.stopPropagation();">Close</Button>
                                </Div>
                                <Div Padding="Padding.Is2" @onclick:stopPropagation="true" Overflow="Overflow.Visible">
                                    <Dropdown Style="z-index:2000;">
                                        <DropdownToggle Size="Size.Small" Clicked="ToggleClicked">
                                            ...
                                        </DropdownToggle>
                                        <DropdownMenu>
                                            <DropdownItem>Add responder</DropdownItem>
                                            <DropdownItem>Delete</DropdownItem>
                                        </DropdownMenu>
                                    </Dropdown>
                                </Div>
                            </Div>
                        </OverlayTemplate>
</DataGrid>
```

The DataGrid only displays this fragment when an OverlayTemplate is present and the mouse is over a row. 
The fragment is removed when the mouse leaves the row.

stather commented:
Added emails to account

stather commented:
recheck

stather commented:
I have read the CLA Document and I hereby sign the CLA

stather commented:
recheck

stsrki commented on Source/Extensions/Blazorise.DataGrid/Blazorise.DataGrid.csproj:
```diff
@@ -9,6 +9,7 @@
   <PropertyGroup>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <NoWarn>1591</NoWarn>
+    <Version>1.1.3</Version>
```
We control versions from a `Blazorise.props` file. You can safely remove this line.

stather replied:
Removed line

stsrki commented on Source/Blazorise/Components/Table/TableRow.razor.cs:
```diff
@@ -111,6 +131,16 @@ public Cursor HoverCursor
+        /// <summary>
+        /// Occurs when the row is mouse overed.
+        /// </summary>
+        [Parameter] public EventCallback<BLMouseEventArgs> MouseOvered { get; set; }
```
Rename to `MouseOver`.

stather replied:
Changed to MouseOver

stsrki commented on Source/Extensions/Blazorise.DataGrid/_DataGridRow.razor:
```diff
@@ -64,6 +66,15 @@
+        @if (ParentDataGrid.OverlayTemplate != null && mouseIsOver)
+        {
+            <TableRowCell Style="width:0px;margin:0px;padding:0px" Overflow="Overflow.Visible">
+                <Div Style="position:absolute;right:0px;margin:0px;padding;0px">
```
Prefer Blazorise utilities over inline styles

- Position
- Margin
- Padding

stather replied:
Changed to use utilities

stsrki commented on .cr/personal/FavoritesList/List.xml:
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
I'm not sure what is `.cr/personal/FavoritesList/List.xml` but consider removing it or add it to a `.gitignore`

stather replied:
Deleted file and added to .gitignore

stsrki commented on Source/Blazorise/Components/Table/TableRow.razor:
```diff
@@ -2,6 +2,8 @@
 @inherits BaseDraggableComponent
 <tr id="@ElementId" class="@ClassNames" style="@StyleNames" @onclick="@OnClickHandler"
     draggable="@DraggableString"
+    @onmouseover="@OnMouseOverHandler"
```
Do we need `:preventDefault` for these events?

stather replied:
It doest seem like it

stather commented:
recheck

stather commented:
@stsrki Hi, I've made the changes. I cannot see for the life of me why the CLA check is failing. Any ideas?

stsrki commented:
`OverlayTemplate` might be a too generic name. My first guess was that it covers the entire of the DataGrid area. Better naming would be a `RowOverlayTemplate`. Any other ideas? 

Also might be a good idea to be able to define overlay position, Start or End of the row.

stather commented:
Renamed parameter to RowOverlayTemplate.
Added Parameter RowOverlayPosition to specify overlay should be positioned at start or end of row.
Added corresponding enum for RowOverlayPosition

stsrki commented on Source/Extensions/Blazorise.DataGrid/Enums/RowOverlayPosition.cs:
```diff
@@ -0,0 +1,18 @@
+        /// Overlay start at left of row.
+        /// </summary>
+
+        Left,
```
Prefer naming `Start` and `End` instead of `Left` and `Right`.

stather replied:
Renamed as requested

stsrki replied:
You're quick :)

stsrki commented:
Hm, it might be. @stather can you check this

![image](https://user-images.githubusercontent.com/900302/203554590-70071ed5-9b64-4f89-9913-99ebd13a6d28.png)

stather commented:
Output from git how seems to have the correct details.
```
C:\Users\rstath698\source\repos\operait\Blazorise>git show
commit 5afdd61d32c3e57a5407c01276ed7aa5ccb5cfa5 (HEAD -> master-datagrid-row-overlay, origin/master-datagrid-row-overlay)
Author: stather <russell.stather@mac.com>
Date:   Wed Nov 23 11:16:34 2022 +0200

    Updated terminology to Start and End
```

David-Moreira commented:
Can you check this one ? Any reason why you closed the PR? Are you redoing it?
![image](https://user-images.githubusercontent.com/22283161/203660085-818a5a5f-1381-4b6c-a522-99c21a2c1fd7.png)
